File: src.csharp/AlphaTab/Core/EcmaScript/DataView.cs
using System;

namespace AlphaTab.Core.EcmaScript
{
    public class DataView
    {
        public ArrayBuffer Buffer { get; }

        public DataView(ArrayBuffer buffer)
        {
            Buffer = buffer;
        }

        public double GetUint8(double offset)
        {
            return Buffer.Raw.Array![Buffer.Raw.Offset + (int) offset];
        }

        public void SetUint16(double offset, double value, bool littleEndian)
        {
            var bytes = BitConverter.GetBytes((ushort) value);
            if (littleEndian != BitConverter.IsLittleEndian)
            {
                System.Array.Reverse(bytes);
            }

            System.Buffer.BlockCopy(bytes, 0, Buffer.Raw.Array!, Buffer.Raw.Offset + (int) offset,
                bytes.Length);
        }

        public double GetInt16(double offset, bool littleEndian)
        {
            var bytes = new byte[sizeof(short)];
            System.Buffer.BlockCopy(Buffer.Raw.Array!, Buffer.Raw.Offset + (int) offset, bytes, 0,
                bytes.Length);
            if (littleEndian != BitConverter.IsLittleEndian)
            {
                System.Array.Reverse(bytes);
            }

            return BitConverter.ToInt16(bytes, 0);
        }

        public void SetInt16(double offset, double value, bool littleEndian)
        {
            var bytes = BitConverter.GetBytes((short) value);
            if (littleEndian != BitConverter.IsLittleEndian)
            {
                System.Array.Reverse(bytes);
            }

            System.Buffer.BlockCopy(bytes, 0, Buffer.Raw.Array!, Buffer.Raw.Offset + (int) offset,
                bytes.Length);
        }

        public double GetUint32(double offset, bool littleEndian)
        {
            var bytes = new byte[sizeof(uint)];
            System.Buffer.BlockCopy(Buffer.Raw.Array!, Buffer.Raw.Offset + (int) offset, bytes, 0,
                bytes.Length);
            if (littleEndian != BitConverter.IsLittleEndian)
            {
                System.Array.Reverse(bytes);
            }

            return BitConverter.ToUInt32(bytes, 0);
        }

        public double GetInt32(double offset, bool littleEndian)
        {
            var bytes = new byte[sizeof(uint)];
            System.Buffer.BlockCopy(Buffer.Raw.Array!, Buffer.Raw.Offset + (int) offset, bytes, 0,
                bytes.Length);
            if (littleEndian != BitConverter.IsLittleEndian)
            {
                System.Array.Reverse(bytes);
            }

            return BitConverter.ToInt32(bytes, 0);
        }

        public void SetInt32(double offset, double value, bool littleEndian)
        {
            var bytes = BitConverter.GetBytes((int) value);
            if (littleEndian != BitConverter.IsLittleEndian)
            {
                System.Array.Reverse(bytes);
            }

            System.Buffer.BlockCopy(bytes, 0, Buffer.Raw.Array!, Buffer.Raw.Offset + (int) offset, bytes
                .Length);
        }

        public double GetUint16(double offset, bool littleEndian)
        {
            var bytes = new byte[sizeof(ushort)];
            System.Buffer.BlockCopy(Buffer.Raw.Array!, Buffer.Raw.Offset + (int) offset, bytes, 0,
                bytes.Length);
            if (littleEndian != BitConverter.IsLittleEndian)
            {
                System.Array.Reverse(bytes);
            }

            return BitConverter.ToUInt16(bytes, 0);
        }

        public void SetUint8(double offset, double value)
        {
            Buffer.Raw.Array![Buffer.Raw.Offset + (int) offset] = (byte) value;
        }

        public double GetInt8(double offset)
        {
            return (sbyte) Buffer.Raw.Array![Buffer.Raw.Offset + (int) offset];
        }

        public double SetUint32(double offset, double value, bool littleEndian)
        {
            var bytes = BitConverter.GetBytes((uint)value);
            if (littleEndian != BitConverter.IsLittleEndian)
            {
                System.Array.Reverse(bytes);
            }

            System.Buffer.BlockCopy(bytes, 0, Buffer.Raw.Array!, Buffer.Raw.Offset + (int) offset, bytes
                .Length);
            return value;
        }
    }
}

File: src.kotlin/alphaTab/src/commonMain/kotlin/alphaTab/core/ecmaScript/DataView.kt
package alphaTab.core.ecmaScript

import alphaTab.core.BitConverter

@ExperimentalUnsignedTypes
class DataView {
    private val _buffer: ArrayBuffer

    public constructor(buffer: ArrayBuffer) {
        this._buffer = buffer
    }

    public fun setUint16(offset: Double, value: Double, littleEndian: Boolean) {
        BitConverter.put(
            _buffer.raw.asByteArray(),
            offset.toInt(),
            value.toUInt().toUShort(),
            littleEndian
        )
    }

    public fun getInt16(offset: Double, littleEndian: Boolean): Double {
        return BitConverter.getInt16(_buffer.raw.asByteArray(), offset.toInt(), littleEndian)
            .toDouble()
    }

    public fun setInt16(offset: Double, value: Double, littleEndian: Boolean) {
        BitConverter.put(
            _buffer.raw.asByteArray(),
            offset.toInt(),
            value.toInt().toShort(),
            littleEndian
        )
    }

    public fun getUint32(offset: Double, littleEndian: Boolean): Double {
        return BitConverter.getUint32(_buffer.raw.asByteArray(), offset.toInt(), littleEndian)
            .toDouble()
    }

    public fun setInt32(offset: Double, value: Double, littleEndian: Boolean) {
        BitConverter.put(_buffer.raw.asByteArray(), offset.toInt(), value.toInt(), littleEndian)
    }

    public fun getUint16(offset: Double, littleEndian: Boolean): Double {
        return BitConverter.getUint16(_buffer.raw.asByteArray(), offset.toInt(), littleEndian)
            .toDouble()
    }

    public fun setUint8(offset: Double, value: Double) {
        _buffer.raw[offset.toInt()] = value.toUInt().toUByte()
    }

    public fun getInt8(offset: Double): Double {
        return _buffer.raw[offset.toInt()].toByte().toDouble()
    }
}
